text/0064-asset-garbage-collection.md

```diff
@@ -0,0 +1,171 @@
+# Garbage Collection for Assets
+
+* **Original Author(s):**: @eladb, @kaizen3031593
+* **Tracking Issue**: #64
+* **API Bar Raiser**: @nlynch
+
+The asset garbage collection CLI will identify and/or delete unused CDK assets,
+resulting in smaller bucket/repo size and less cost for customers.
+
+## Working Backwards
+
+**CHANGELOG**:
+
+- feat(cli): `cdk gc` - garbage collection of unused assets
+
+**Help**:
+
+```shell
+➜ cdk gc --help
+cdk gc [ENVIRONMENT...]
+
+Finds and deletes all unused S3 and ECR assets in the ENVIRONMENT
+
+Options:
+  -l, --list              lists, rather than deletes, all unused assets
+  -t, --type=[s3|ecr]     filters for type of asset
+
+Examples:
+  cdk gc
```

Comment:
Can we make this zero-touch?

Could it be something we install into the account that runs periodically? I think that should be the end goal of the user experience.

We could complement that with a CLI experience, but only if one leads naturally into the other (in terms of development effort).

Reply:
I could be way off base but a zero-touch solution would likely be a lambda function that runs once a week (or however long) versus the CLI experience which is by manual command. But at the end of the day, the actual code that gets run (tracing the assets in the environment and then deleting the unused ones) should be exactly the same.

It seems like if the CLI experience works then integrating it into a zero-touch solution would not require much additional work.

Reply:
Yes, +1 on @kaizen3031593's response - I agree that eventually a scheduled tasks runs GCs is what users would want but I also think it makes sense to start by implementing `cdk gc` as the foundational building block that will be used in a automated task.

After we have a solid manual `cdk gc` (which will take some time to stabilize), we (or the community) can vend a construct that runs this periodically and maybe at some point we can include that in our bootstrapping stack.

Reply:
It would be interesting to be able to opt-in to this behavior as part of the bootstrap process. The `CdkToolkit` stack could maybe be the owner of the scheduled Lambda function that runs the GC process.